src/common/utils/method.utils.ts

```diff
@@ -15,24 +15,14 @@ export function normalizeArray<T>(
   return normalizeObj;
 }
 
-export function groupBy<T>(
+export function groupBy<T, K extends keyof any>(
   arr: T[],
-  selector: (item: T) => string | number | null = (item: any) => item.id,
-) {
-  const map: Record<string, T[] | undefined> = {};
-
-  arr.forEach((data) => {
-    const key = selector(data);
-    if (key !== null) {
-      if (map[key]) {
-        map[key]!.push(data);
-      } else {
-        map[key] = [data];
-      }
-    }
-  });
-
-  return map;
+  key: (i: T) => K,
+): Record<K, T[]> {
+  return arr.reduce((groups, item) => {
+    (groups[key(item)] ||= []).push(item);
+    return groups;
+  }, {} as Record<K, T[]>);
```

Comment:
```suggestion
  selector: (i: T) => K = (i any) => i.id,
): Record<K, T[] | undefined> {
  return arr.reduce((groups, item) => {
    (groups[key(item)] ??= []).push(item);
    return groups;
  }, {} as Record<K, T[]> | undefined);
```

1. 두 번째 파라미터는, return으로 key를 가져오는 함수니까, key보다는 selector가 더 어울리지 않을까요?
2. 기존에 selector 파라미터에 default값이 있어서 추가해야하지 않을까요? (쓰는 곳이 없긴 하지만)
3. OR보다 nullish가 더 맞지 않을까요
4. 그리고 여기서 default로 []를 넣어준 것 같은데, Record에 접근한 key가 없을 때 undefined를 뱉는건 똑같아서 return 타입에 undefined를 or로 명시하자는 말이었어요.

Reply:
1. 동의합니다.
2. 있어서 제거했습니다. [참고](https://github.com/sparcs-kaist/otlplus-server/pull/46/commits/1e2f031111be71e8f1469e8807a7188ef3e8fea1#diff-ba78ab18c57f73d011885c8d42d59dc8fb815f31022b05a35bde2f0bea57960eL153-R157)
3. 동의합니다. 코드 긁어와서 읽었는데 그 부분은 생각을 못했네요
4. 동의합니다. 다만 실사용시에는 Dto["tag"] 이런 느낌으로 arbitrary string 대신 string list등 type을 더 명확히 사용하면 undefined를 정확히 사용할 수 있지 않을까 싶기도 합니다. (근데 이러면 type system이 아니라 그냥 assertion 되는거라 undefined 넣는게 어쩔 수 없는 것 같기도...?)